---
title: "NF-YC gene family case study"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
bibliography: synder.bib 
output:
    rmarkdown::html_vignette:
        fig_caption: yes 
vignette: >
  %\VignetteIndexEntry{NFYC case study}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

This is an internal vignette that uses local data. I'll make the data
accessible soon, but for now you shouldn't use this vignette.

```{r, eval=FALSE}
require(ape)
tree <- read.tree('arabidopsis/tree')
pdf('tree.pdf')
plot(tree)
dev.off()
```

```{r, eval=FALSE}
collect_NFYC_data <- function(
    syn_file,            # syn_file="arabidopsis/syn/at-vs-al.syn"
    fgff_file,           # fgff_file="arabidopsis/gff/yc.gff"
    tgff_file,           # tgff_file="arabidopsis/gff/al.gff"
    gene_map_file,       # gene_map_file="arabidopsis/yc.tab"
    tblastn_file,        # tblastn_file="arabidopsis/al.blast.tab"
    blastp_file = NULL,  # blastp_file="arabidopsis/yc-vs-al.blast.tab"
    trans  = "d",        # trans="d"
    k      = 0L,         # k=0L
    r      = 0           # r=0
){
  extractNameFromAttr <- function(x){
    sub(".*Name=([^;]+).*", "\\1", x)
  }

  # load synteny map
  syn  <- synder::as_synmap(syn_file)

  # load focal GFF
  fgff <- .tidy_gff(fgff_file, prefix="f", fromAttr=extractNameFromAttr)

  # load the transcribed intervals target GFF
  tgff <- .tidy_gff(tgff_file, prefix="t", fromAttr=extractNameFromAttr)
  # FIXME: do I have to restrict this to mRNA?
  tgff <- tgff[tgff$ttype == "mRNA", ]

  # load map with name (NF-YC*), gene (locus ID), to transcript (model ID)
  nfyc <- read.table(
    gene_map_file,
    col.names=c("fgene_name", "fseqid", "fprotein"),
    stringsAsFactors=FALSE
  )
  # a map from fprot_id to fseqid
  fmap <- nfyc[, c(3,2)]

  # map focal genes to target search intervals 
  srcres <- synder::search(syn, synder::as_gff(fgff_file), k=k,r=r,trans=trans)
  srcres <- .tidy_searchResult(srcres, fromAttr=extractNameFromAttr)

  tblastn_si_map <- synder::make_tblastn_si_map(tblastn_file, srcres, fmap=fmap)
  tblastn_gene_map <- synder::make_tblastn_gene_map(tblastn_file, tgff)

  if(is.null(blastp_file)){
    blastp_raw <- NULL 
    blastp_map <- NULL
  } else {
    x <- make_blastp_map(
      blastp_file=blastp_file,
      tgff=tgff,
      fgff=fgff,
      srcres=srcres,
      fmap=fmap
    )
    blastp_raw <- x$blastraw
    blastp_map <- x$blastmap
  }

  feats <- featureMap(srcres=srcres, fgff=fgff, tgff=tgff)

  tblastn_raw <- tblastn_si_map$blastraw
  tblastn_si_map <- tblastn_si_map$blastmap
  tblastn_gene_map <- tblastn_gene_map$blastmap

  .namedlist(syn, tgff, fgff, nfyc, srcres, tblastn_raw, tblastn_si_map, tblastn_gene_map, blastp_raw, blastp_map, feats)
}

files <- list(
  al=list(syn="syn/at-vs-al.syn", gff="gff/al.gff", tblastn="al.blast.tab", blastp="yc-vs-al.blast.tab"),
  cr=list(syn="syn/at-vs-cr.syn", gff="gff/cr.gff", tblastn="cr.blast.tab"),
  br=list(syn="syn/at-vs-br.syn", gff="gff/br.gff", tblastn="br.blast.tab"),
  es=list(syn="syn/at-vs-es.syn", gff="gff/es.gff", tblastn="es.blast.tab")
)

results <- lapply(files, function(x){
  collect_NFYC_data(
    fgff_file     = "arabidopsis/gff/yc.gff",
    gene_map_file = "arabidopsis/yc.tab",
    syn_file      = paste0("arabidopsis/", x$syn),
    tgff_file     = paste0("arabidopsis/", x$gff),
    tblastn_file  = paste0("arabidopsis/", x$tblastn),
    blastp_file   = if(!is.null(x$blastp)){paste0("arabidopsis/", x$blastp)}else{NULL},
    k=0L, r=0, trans="d"
  )
})
```

```{r}
results <- loadRDS('results.rds')

tabulate_from <- function(xs, key){
  ys <- lapply(xs, function(x) x[[key]])
  for(grp in names(ys)){
    ys[[grp]]$group <- grp
  }
  do.call(rbind, ys) %>% .remove_rownames
}

nuniq <- function(x){ length(unique(x)) }

show_vec <- function(xs, nmax=5) {
  if(nuniq(xs) <= nmax) {
    paste(unique(xs), collapse=",")
  } else {
    "lots"
  }
}

### ===========================================================================
# results based solely on synder
feature_maps <- tabulate_from(results, "feats")
dplyr::group_by(feature_maps, group, fseqid) %>%
  dplyr::summarize(ntargets = nuniq(tseqid), targets=show_vec(tseqid, 5))


### ===========================================================================
# results based solely on tblastn
tblastns <- tabulate_from(results, "tblastn_raw") %>%
  dplyr::filter(tn_evalue < 0.001)

# get the total number of tBLASTn hits for the model with the most number of hits from each focal gene
# NOTE: this will include exons
total_tblastn_hits_by_gene <- tblastns %>%
  dplyr::group_by(group, fprotein) %>%
  dplyr::summarize(hits = length(fprotein), fseqid=fseqid[1]) %>%
  dplyr::group_by(group, fseqid) %>%
  dplyr::summarize(max_hits = max(hits))
# TODO: merge the blast results with the target GFF, merge hits to the same interval
# FIXME: do I need to consider frame?

### ===========================================================================
tblastn_si_maps <-  tabulate_from(results, "tblastn_si_map")
tblastn_si_maps$cset <- as.character(tblastn_si_maps$cset)
tblastn_si_maps <- dplyr::filter(tblastn_si_maps, tn_evalue < 0.001)

# get the total number of tBLASTn hits for the model with the most number of
# hits from each focal gene where the gene blasted and syndered are the same
# NOTE: the number of hits includes hits to different exons
dplyr::filter(tblastn_si_maps, fseqid == si_fseqid) %>%
  dplyr::group_by(group, fprotein) %>%
  dplyr::summarize(hits = length(fprotein), fseqid=fseqid[1]) %>%
  dplyr::group_by(group, fseqid) %>%
  dplyr::summarize(max_hits = max(hits))

# Find overlaps with tgff protein features, this resoves the above exon issue
# FIXME: do I need to consider frame?
tgffs <- tabulate_from(results, "tgff")
tgffs$group=NULL
.overlaps(
  x=tblastn_si_maps, xid="tchr", xa="tn_tstart", xb="tn_tstop",
  y=tgffs, yid="tchr", ya="tstart", yb="tstop"
) %>%
  # extract rows where the protein coded for by query gene i has a tBLASTn hit
  # within a search interval inferred for the same gene. 
  dplyr::filter(fseqid == si_fseqid) %>%
  dplyr::group_by(group, fseqid) %>%
  dplyr::summarize(ntargets = nuniq(tseqid), targets=show_vec(tseqid, 9))

# summarize each group
dplyr::group_by(tblastn_si_maps, group) %>%
  dplyr::summarize(
    fseqids = length(fseqid),
    sets = nuniq(cset),
    agreements = mean(fseqid == si_fseqid),
    tchrs = nuniq(tchr)
  )

# summarize each cset
dplyr::group_by(tblastn_si_maps, cset) %>%
  dplyr::summarize(
    fseqids = nuniq(fseqid),
    tseqids = nuniq(si_fseqid),
    min_eval = min(tn_evalue),
    fseqids_ref = show_vec(fseqid),
    si_fseqids = show_vec(si_fseqid)
  )
dplyr::group_by(tblastn_si_maps, group, si_fseqid) %>%
  dplyr::summarize(csets = nuniq(cset))


### ===========================================================================
# FIXME: why do I even make this table? can I nuke it?
tblastn_gene_maps <- tabulate_from(results, "tblastn_gene_map")
tblastn_gene_maps <- dplyr::filter(tblastn_gene_maps, tn_evalue < 0.001)

head(tblastn_gene_maps)
dplyr::group_by(tblastn_gene_maps, group, fseqid) %>%
  dplyr::summarize(nfeat = nuniq(feat_tname))

```
